### DISCUSSION TITLE
Bring rustreexo into rust-bitcoin org

### DISCUSSION THREAD

**Discussion by apoelstra** [rust-bitcoin/rust-bitcoin#1318] [Other Bitcoin-related crates]
I would like to bring MIT DCI's utreexo library into the rust-bitcoin org (and @kcalvinalvin asked me about this at a conference). There are a few small technical things that need to be improved:

* drop the `sha2` dependency (there is already a dep on `bitcoin_hashes` so it should be easy to just switch over to that, though there'll be a perf hit unfortunately
* update the `bitcoin` and `bitcoin_hashes` dependency
* bring MSRV down to 1.41 and check this in CI (there are a couple minor syntax things that'll have to be changed but nothing serious)

This library is actively maintained by the DCI, which is a trustworthy and morally aligned organization, and I really like utreexo as a project. It is not part of the bitcoin p2p protocol (yet?) so is arguably off-topic, but it is a proposal *for* bitcoin. And deployment would not actually require any fork/consensus from the network, just for some people to run compatible software.

What do people think?

cc @kcalvinalvin @Davidson-Souza

**Comment by apoelstra** [comment]
Oops, I forgot the link: https://github.com/mit-dci/rustreexo/

**Comment by sanket1729** [comment]
concept ACK. Utreexo is a promising project. I personally know Calvin and can vouch for him.

**Comment by Davidson-Souza** [comment]
As one of utreexo's contributor, I like the idea. Being in an org like rust-bitcoin can bring more visibility and reviews to our lib.  
As for the required changes, they are in my to-do list already, and won't be a big obstructor.  

> It is not part of the bitcoin p2p protocol (yet?) so is arguably off-topic  

Yes, it's not part of the p2p protocol yet, but we are working on it (I have it partially working on my rust-bitcoin fork).

**Comment by sanket1729** [comment]
Apart from the above changes, other contributors may demand some code quality changes before bringing it in or releasing a 0.1.0. From my side, having a functioning CI with a (decent) test coverage seems is a must-have. I also see lots of commented code, perhaps the remains of some debugging in some files. Would be good to clean these up too. 

Would have more detailed feedback once the repo is ready for consideration.

**Reply by kcalvinalvin** [reply]
Sounds good. We have a decent test coverage already with the rustreexo repo and are looking to add fuzz testing as well.

We'll work towards a 0.1.0 and ping again once the repository is at that stage :)

**Reply by Kixunil** [reply]
I wouldn't mind having the repo even before cleaning up if it has a visible "WIP" warning in README.

**Comment by Kixunil** [comment]
"Digital currency initiative" doesn't sound appealing but if you know them and say they are OK, I trust you on this.

**Reply by apoelstra** [reply]
Their mandate is to research various things in the space and produce both academic papers and reports for institutional/regulatory bodies. They also produce novel cryptographic research (e.g. they employ at least one author of the groundbreaking "SNARKs for C" paper, as well as the primary author of the utreexo proposal; in the past they've employed LN developers, including one of the original authors, as well as Bitcoin Core developers.)

This does mean investigating all sorts of unsavory projects, but they don't wade into it and are not associated with any professorcoins or other scams, even ones from MIT.

**Comment by kcalvinalvin** [comment]
> * drop the `sha2` dependency (there is already a dep on `bitcoin_hashes` so it should be easy to just switch over to that, though there'll be a perf hit unfortunately

While working on this I found it's a bit tricky to drop the sha2 dependency for now. The hash function we're using in utreexo is SHA512_256 and we've switched to it from SHA256 as the bigger block size benchmarked to be faster. Since bitcoin_hashes doesn't have SHA512_256, we can't drop it unless bitcoin_hashes is changed to support SHA512_256.

I will note that the hash function used in utreexo isn't final and it is subject to change (to whatever is the fastest hash function that's palatable by Bitcoin Core people). Though since every other utreexo library uses SHA512_256, I'd prefer to keep compatibility and keep the sha2 dependency for now.

**Reply by Kixunil** [reply]
The opposition against `sha2` is pretty strong here. I myself wouldn't mind an exception with the clear intention to change it in the future.

**Reply by apoelstra** [reply]
What is SHA512_256?

**Reply by apoelstra** [reply]
If it is just truncated sha512 we can certainly support that. It looks like maybe it has a different IV too?

**Reply by elichai** [reply]
Have you actually benchmarked them? Last time I tried I didn't notice SHA512_256 to be significantly better. 
You should also check how SHA3_256 performs, as it's already in bitcoin core (for TORv3 support)

**Reply by kcalvinalvin** [reply]
> If it is just truncated sha512 we can certainly support that. It looks like maybe it has a different IV too?

iirc it's truncated sha512 but with different constants so that a truncated sha512 and sha512_256 outputs are different.

**Reply by apoelstra** [reply]
Cool, we'll add it once we have #1284 in.

**Reply by kcalvinalvin** [reply]
> Have you actually benchmarked them? Last time I tried I didn't notice SHA512_256 to be significantly better.

Yes. The relevant PR is:
https://github.com/mit-dci/utreexo/pull/193

I didn't comment on there but I independently verified it at the time.

iirc it wasn't because sha512 is faster than sha256 but because the leaf preimages that we're hashing are at least 928 bits. So with the bigger chunk sizes in sha512 there was less looping when hashing the leaves.

> You should also check how SHA3_256 performs, as it's already in bitcoin core (for TORv3 support)

Ah wasn't aware. I'm definitely open to using sha3. Are there plans to support it in rust-bitcoin?

**Reply by apoelstra** [reply]
We could also support sha3, yes, though I'd need some specific usecase for it.

**Reply by elichai** [reply]
@kcalvinalvin Note that the go implementaion could behave *very* differently than the rust ones :) (even though their crypto code is usually pretty good, especially the `Sum` function, the compiler is much less optimizing)

**Reply by kcalvinalvin** [reply]
> @kcalvinalvin Note that the go implementaion could behave _very_ differently than the rust ones :) (even though their crypto code is usually pretty good, especially the `Sum` function, the compiler is much less optimizing)

Different language implementations differing in performance is a good point I haven't thought of before. I'll put that on the todo list.

My initial impression was that sha512_256 should at least consistently be faster than sha256 for us across different languages but yeah there's on data to back that.

**Comment by elichai** [comment]
Concept ACK here too, utreexo is pretty cool and I think it'll be better for the project to use the same conservative guidelines as rust-bitcoin does

**Comment by Davidson-Souza** [comment]
Pinging this discussion after releasing [v0.1.0](https://github.com/mit-dci/rustreexo/releases/tag/v0.1.0). Since the original post, we dropped the `sha2` dependency, we don't use `bitcoin` but `bitcoin-hases` and we are on a newer version (it's easy to bump to the latest if required). The MSRV is `1.41` and we have a CI to test on this version. There's no unused code anymore. The API is entirely documented, with some useful examples included.

`rustreexo` also is linted with `cargo fmt` and `cargo clippy`, this is CI checked too. It also implements [de]-serialization with both a custom one or using `serde` with a feature `with-serde`. We have CSNs, bridge node, and full Proof support. There are more structures being developed, but this is already enough to build full `utreexo` applications. We have a [crates.io](https://crates.io/crates/rustreexo/) package and ffi bindings are [on the way](https://github.com/mit-dci/rustreexo/pull/37). 

Any feedback and review for the lib are very welcome!

**Reply by apoelstra** [reply]
This looks great! I quickly skimmed the code, which looks very clean and didn't have anything scary like unsafe code or filesystem access (outside of reading some test files). It has a very low dependency footprint, pretty-much just serde, which is feature-gated.

I think this is ready to come into the org. Though I'll give others a week or two to express any concerns.

**Reply by tcharding** [reply]
I don't see anything to prevent it coming into the org. The code is clean, there are a few un-rusty things and a few code style things that are different (not that every repo in the org needs the same coding style). Must have been a fun project to work on, pure data structure stuff, I bet you know more about `RefCell` than the average Rust dev now.

**Reply by sanket1729** [reply]
We can bring this in. I don't see any issues with the codebase now. Out of curiosity, have you tested these test vectors with implementations of utreexo?

**Reply by Davidson-Souza** [reply]
@tcharding 

> there are a few un-rusty things and a few code style things that are different 

I don't mind code style changes, I welcome feedbacks on that too. I'm not the most experienced Rust dev you'll find out there, but I think it's nice to have a code following a well-defined styling.

> Must have been a fun project to work on, pure data structure stuff, I bet you know more about RefCell than the average Rust dev now.

Yes! I had to write the Pollard thing for scratch three times, to make it work. I've learned a lot about Rust memory management.

@sanket1729 
> Out of curiosity, have you tested these test vectors with implementations of utreexo?

All test cases in the json files are generated from [the go lib](https://github.com/utreexo/utreexo). I also plan to make a cross-fuzzing between rustreexo and the utreexo golang lib.
I'm also using rustreexo in a [project](https://github.com/Davidson-Souza/floresta), and didn't find any incompatibilities while ingesting proofs from [utreexod](https://github.com/utreexo/utreexod) into Floresta.

I have one question, though. We are planing into making language bindings to rustreexo. I already have C and Python bindings working for Stump and Proof. Would it be better to keep those bindings somewhere else, or in the main repo?

**Reply by sanket1729** [reply]
Awesome.

> Would it be better to keep those bindings somewhere else, or in the main repo?

We have never had bindings from rust to other languages in any rust-bitcoin repos. I don't any reason as to why not, completely your decision. Please note that I have not written bindings from rust to other languages to form an informed opinion. I would hope that bindings do not incur any additional run-time dependencies for regular rust library users.

**Reply by apoelstra** [reply]
I've done a few small-scale rust bindings to C before. C at least does not require anything special, you just have to write some functions that have a limited interface and some #[extern] declaration or something.

**Comment by yancyribbens** [comment]
Nice!  utreexo is a cool project I think and I'm glad to see there's a Rust implementation.

**Comment by stevenroose** [comment]
LGTM

**Comment by jaoleal** [comment]
So, when raising the [floresta](https://github.com/vinteumorg/Floresta/pull/257) problem of using a fork of rust-bitcoin maintained by @Davidson-Souza and trying to resolve it, I found two starting points for implementing rustreexo support in this lib :

* rustreexo itself 

* p2p messaging (A sketch done by davidson https://github.com/Davidson-Souza/rust-bitcoin/tree/patch/src/network )

How is your (maintainers) mood regarding this? 😀

**Reply by tcharding** [reply]
I have zero knowledge on the topic, happy to help review.

**Reply by jaoleal** [reply]
Superficially, utreexo gets the utxo state and stores it in a tree, and you can validate things on-chain with only that tree, log(n) to store the state. its actually a lot of trees thats why the project floresta has its name, the brazilian mean for forest... 

kinda obvious last part but i really like it :laughing:
